chain/rosetta-rpc/src/adapters/mod.rs

```diff
@@ -144,54 +142,66 @@ pub(crate) async fn convert_block_to_transactions(
         .await?
         .map_err(crate::errors::ErrorKind::InternalError)?;
 
-    let mut transactions = Vec::<crate::models::Transaction>::new();
-    for account_change in accounts_changes.into_iter() {
+    let transactions = convert_block_changes_to_transactions(
+        &genesis.config.runtime_config,
+        &block.header.hash,
+        accounts_changes,
+        accounts_previous_state,
+    )?;
+    Ok(transactions.into_iter().map(|(_transaction_hash, transaction)| transaction).collect())
+}
+
+fn convert_block_changes_to_transactions(
+    runtime_config: &near_runtime_configs::RuntimeConfig,
+    block_hash: &near_primitives::hash::CryptoHash,
+    accounts_changes: near_primitives::views::StateChangesView,
+    mut accounts_previous_state: std::collections::HashMap<
+        near_primitives::types::AccountId,
+        near_primitives::views::AccountView,
+    >,
+) -> Result<std::collections::HashMap<String, crate::models::Transaction>, crate::errors::ErrorKind>
+{
+    use near_primitives::views::StateChangeCauseView;
+
+    let mut transactions = std::collections::HashMap::<String, crate::models::Transaction>::new();
+    for account_change in accounts_changes {
         let transaction_hash = match account_change.cause {
-            near_primitives::views::StateChangeCauseView::TransactionProcessing { tx_hash } => {
+            StateChangeCauseView::TransactionProcessing { tx_hash } => {
                 format!("tx:{}", tx_hash.to_base())
             }
-            near_primitives::views::StateChangeCauseView::ActionReceiptProcessingStarted {
-                receipt_hash,
-            } => format!("receipt:{}", receipt_hash.to_base()),
-            near_primitives::views::StateChangeCauseView::ActionReceiptGasReward {
-                receipt_hash,
-            } => format!("receipt:{}", receipt_hash.to_base()),
-            near_primitives::views::StateChangeCauseView::ReceiptProcessing { receipt_hash } => {
+            StateChangeCauseView::ActionReceiptProcessingStarted { receipt_hash }
+            | StateChangeCauseView::ActionReceiptGasReward { receipt_hash }
+            | StateChangeCauseView::ReceiptProcessing { receipt_hash }
+            | StateChangeCauseView::PostponedReceipt { receipt_hash } => {
                 format!("receipt:{}", receipt_hash.to_base())
             }
-            near_primitives::views::StateChangeCauseView::PostponedReceipt { receipt_hash } => {
-                format!("receipt:{}", receipt_hash.to_base())
+            StateChangeCauseView::InitialState => format!("block:{}", block_hash),
+            StateChangeCauseView::ValidatorAccountsUpdate => {
+                format!("block-validators-update:{}", block_hash)
```

Comment:
doesn't this mean that the transaction identifier will be the same for all validator account updates? Is that intended?

Reply:
Good question! Well, Rosetta is fine doing it this way. The problem is that from NEAR point of view there cannot be a transaction to multiple receiver ids (in Rosetta it is possible to express), so you cannot reconstruct such a transaction, but given it is a "fake transaction" anyway, we can get away with it, I guess. Another problematic endpoint is the first block, which we construct from genesis records (and it is a massive single fake transaction with tons of operations).

Ultimately, I consider this solution as good enough to address the original issue of duplicate transaction identifiers. We may redesign the transaction split with more granular transaction identifiers once we have a specific request for that (from Rosetta point of view, it should be indifferent whether it is a single transaction with a generic transaction identifier and N operations or N transactions with specific transaction identifier and 1 operation in every one of them)